From 616f6334e99fdabeb82aad4e249eaa6c37ca095a Mon Sep 17 00:00:00 2001
From: Jaegeuk Kim <jaegeuk@google.com>
Date: Fri, 20 Mar 2026 03:00:44 +0000
Subject: [PATCH 51/63] ANDROID: mm: add vendor hook in page_cache_ra_order

This vendor hook allows retrying ra_alloc_folio().

Bug: 448150994
Bug: 498623919
Change-Id: I61cd717bdbc925088e287f0cab116c42890885ab
Signed-off-by: Jaegeuk Kim <jaegeuk@google.com>
(cherry picked from commit ebe70b434406e800ea085d9543db91b3797fe28a)
---
 drivers/android/vendor_hooks.c | 1 +
 include/trace/hooks/mm.h       | 3 +++
 mm/readahead.c                 | 9 ++++++++-
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index a8b69d039928..a06eda6682b1 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -437,6 +437,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_lruvec_del_folio);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_async_mmap_readahead);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_free_page);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_page_cache_ra_unbounded);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_ra_alloc_retry);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_force_page_cache_ra);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_fault_post_folio_locked);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_filemap_fault_folio_locked);
diff --git a/include/trace/hooks/mm.h b/include/trace/hooks/mm.h
index 02105aec0a88..7e5a8410b720 100644
--- a/include/trace/hooks/mm.h
+++ b/include/trace/hooks/mm.h
@@ -363,6 +363,9 @@ DECLARE_HOOK(android_vh_mm_free_page,
 DECLARE_HOOK(android_vh_page_cache_ra_unbounded,
 	TP_PROTO(struct address_space *mapping, struct folio *folio, u64 *data),
 	TP_ARGS(mapping, folio, data));
+DECLARE_HOOK(android_vh_ra_alloc_retry,
+	TP_PROTO(unsigned int *order, bool *retry),
+	TP_ARGS(order, retry));
 DECLARE_HOOK(android_vh_force_page_cache_ra,
 	TP_PROTO(struct address_space *mapping, u64 *data),
 	TP_ARGS(mapping, data));
diff --git a/mm/readahead.c b/mm/readahead.c
index 6a0d9d6fba01..9f928fe41359 100644
--- a/mm/readahead.c
+++ b/mm/readahead.c
@@ -543,9 +543,16 @@ void page_cache_ra_order(struct readahead_control *ractl,
 		/* Don't allocate pages past EOF */
 		while (order > min_order && index + (1UL << order) - 1 > limit)
 			order--;
+retry:
 		err = ra_alloc_folio(ractl, index, mark, order, gfp);
-		if (err)
+		if (err) {
+			bool retry = false;
+
+			trace_android_vh_ra_alloc_retry(&order, &retry);
+			if (retry)
+				goto retry;
 			break;
+		}
 		index += 1UL << order;
 	}
 
-- 
2.50.1

